test/test_torch.py

```diff
@@ -6088,6 +6088,101 @@ def test_pow(self, device):
+        exp[0] = exp[4] = exp[6] = 0
+
+        expected = torch.from_numpy(np.float_power(to_np(base), to_np(exp)))
+
+        for op in (torch.float_power, torch.Tensor.float_power, torch.Tensor.float_power_):
+            result = op(base.clone(), exp)
+            self.assertEqual(expected, result)
+
+            if op is torch.float_power:
+                out = torch.empty_like(base).to(device=device, dtype=torch.float64)
+                op(base, exp, out=out)
+                self.assertEqual(expected, out)
+
+            for i in range(10):
+                expected_scalar_exp = np.float_power(to_np(base), i)
+                expected_scalar_base = np.float_power(i, to_np(exp))
+
+                result = op(base.clone(), i)
+                self.assertEqual(expected_scalar_exp, result)
+
+                if op is torch.float_power:
+                    result = op(i, exp)
+                    self.assertEqual(expected_scalar_base, result)
+
+                    out = torch.empty_like(base).to(device=device, dtype=torch.float64)
+                    op(base, i, out=out)
+                    self.assertEqual(expected_scalar_exp, out)
+
+                    out = torch.empty_like(exp).to(device=device, dtype=torch.float64)
+                    op(i, exp, out=out)
+                    self.assertEqual(expected_scalar_base, out)
+
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @onlyOnCPUAndCUDA
+    @dtypes(*(torch.testing.get_all_dtypes(include_bool=False)))
+    def test_float_power_with_complex(self, device, dtype):
+        def to_np(value):
+            if isinstance(value, torch.Tensor):
+                if value.dtype == torch.bfloat16:
+                    value = value.to(dtype=torch.float)
+                return value.cpu().numpy()
+            return value
+
+        base = make_tensor((8,), device, dtype, low=1, high=10)
+        exps = [torch.tensor([-2.8, -2, -1, -0.5, 0, 0.5, 1, 2], device=device),
+                torch.tensor([-2.5j, -1.0j, 0j, 1.0j, 2.5j, 1.0 + 1.0j, -1.0 - 1.5j, 3.3j], device=device)]
+
+        for exp in exps:
+            if dtype.is_complex or exp.dtype.is_complex:
+                expected = torch.from_numpy(np.float_power(to_np(base), to_np(exp)))
+
+                for op in (torch.float_power, torch.Tensor.float_power, torch.Tensor.float_power_):
+                    result = op(base.clone(), exp)
+                    self.assertEqual(expected, result)
+
+                    if op is torch.float_power:
+                        out = torch.empty_like(base).to(device=device, dtype=torch.complex128)
+                        op(base, exp, out=out)
+                        self.assertEqual(expected, out)
+
+    @onlyOnCPUAndCUDA
+    @dtypes(*(torch.testing.get_all_dtypes(include_bool=False)))
+    def test_float_power_exceptions(self, device, dtype):
+        values = [torch.tensor([-2.8, -2, -1, -0.5, 0, 0.5, 1, 2], device=device),
```

Comment:
I think this test can be simplified by writing a helper function for the type promotion:

```
_promo_helper(x, y):
  ...
```

That works with both tensors and scalars and returns the required dtype for the out argument (either torch.double or torch.cdouble). 

Then you can just enumerate a view a/b pairs and try passing the same three outs to each one. For example:

```
test_cases = ((torch.tensor(5., device=device), -.25),
  ...
for a, b in test_cases:
  for out_dtype in (torch.long, torch.float, torch.double, torch.cdouble):
    out = torch.empty(1, device=device, dtype=out_dtype)
    required_dtype = _promo_helper(a, b)
   if out.dtype == required_dtype:
     torch.float_power(a, b, out=out)
   else:
      with self.assertRaisesRegex(RuntimeError, "is not the desired output type"):
        torch.float_power(a, b, out=out)
```

Reply:
Also test `torch.float_power`, like the above, and `torch.Tensor.float_power_`. The latter operation should also require that the first (base) tensor's dtype is the same as the required_dtype, or it will throw its own error.

Reply:
Addressed! Thanks for advice!